File: functions/CopBET_LEiDA_transition_entropy.m
% out = CopBET_LEiDA_transition_entropy(in,K,varargin)
%
% Copenhagen Brain Entropy Toolbox: LEiDA transition entropy.
% Evaluates the leading eigenvectors of instantaneous phase coherence
% matrices, clusters the data using k-means with squared euclidean distance
% function, and evaluates the Markov transition rate entropy, all according
% to the specifications in Kringelbach et al., 2020: "Dynamic coupling of
% whole-brain neuronal and neurotransmitter systems" and the associated
% Github repository.
%
% Input:
%   in: a matrix (nxp,n>1) or a table where the first column contains
%   matrices (in cells) to be concatenated before clustering, e.g.,
%   different subjects or scan sessions.
%   K: number of clusters for k-means clustering. 
% name-value pairs:
%   keepdata: Indicates whether the output table also should contain the
%   input data, i.e., by adding an extra column containing entropy values.
%   Defaults to true
%   parallel: Whether to run 200 replicates of k-means in parallel [true]
%
%
% Neurobiology Research Unit, 2023
% Please cite McCulloch, Olsen et al., 2023: "Navigating Chaos in
% Psychedelic Neuroimaging: A Rigorous Empirical Evaluation of the Entropic
% Brain Hypothesis" if you use CopBET in your studies. Please read the
% paper to get a notion of our recommendations regarding the use of the
% specific methodologies in the toolbox.

% ASO 9/3-2023

% potential tests:
% Check that cluster centroids make sense (qualitatively) and correct size
% Check unit norm of data points
% Check nans/infs

function out = CopBET_LEiDA_transition_entropy(in,K,varargin)

if nargin<2
    error('Please specify a number of components, K')
end

[out,numworkers,in] = CopBET_function_init(in,varargin);

% Do LEiDA and concatenate data for clustering
disp('Running LEiDA')
datasizes = nan(height(in),1);
for ses = 1:height(in)
    datasizes(ses) = size(in{ses,1}{1},1);
end
Lead_eigs_all = nan(sum(datasizes),size(in{1,1}{1},2));
c = 1;
for ses = 1:height(in)
    tmp = in{ses,1}{1};
    tmp = tmp - mean(tmp);
    Lead_eigs_all(c:c+datasizes(ses)-1,:) = LEiDA(tmp)';
    sensible_data_check(Lead_eigs_all(c:c+datasizes(ses)-1,:),['Leading eigenvectors, session',num2str(ses)]);
    c = c+datasizes(ses);
end


disp('Running k-means')
if numworkers>0
    a = gcp;
    if isempty(a)
        parpool(numworkers)
    end
    [IDX,C]=kmeans(Lead_eigs_all,K,'Distance','sqeuclidean',...
        'Replicates',200,'Display','final','Options',statset('UseParallel',1));
else
    [IDX,C]=kmeans(Lead_eigs_all,K,'Distance','sqeuclidean',...
        'Replicates',200,'Display','final');
end

entropy = tpm_entropy(IDX,datasizes);
sensible_data_check([entropy{:}],'entropy');
out.entropy = [entropy{:}]';



end

%% functions
function Leading_Eig = LEiDA(ts)
% More or less copy-pasted from https://github.com/decolab/pnas-neuromod/blob/master/LEiDA_PsiloData.m
% This function can be optimized heavily, however, we tried to keep it as
% much as the original to prevent differences from the original paper.



% fnq=1/(2*TR);                 % Nyquist frequency
% flp = 0.04;                    % lowpass frequency of filter (Hz)
% fhi = 0.07;                    % highpass
% Wn=[flp/fnq fhi/fnq];         % butterworth bandpass non-dimensional frequency
% k=2;                          % 2nd order butterworth filter
% [bfilt,afilt]=butter(k,Wn);   % construct the filter

[Tmax,N_areas] = size(ts);
Phase_BOLD=zeros(N_areas,Tmax); %OBS transposed
%%%%%%%%%%%%% from PNAS code
for seed=1:N_areas
    tstmp=detrend(ts(:,seed));
%     signal_filt =filtfilt(bfilt,afilt,tstmp);
%     Phase_BOLD(seed,:) = angle(hilbert(signal_filt));
    Phase_BOLD(seed,:) = angle(hilbert(tstmp));
    % Phase_BOLD = angle(hilbert(ts))';
end
Leading_Eig = nan(N_areas,Tmax);

for t=1:Tmax
    
    %Calculate the Instantaneous FC (BOLD Phase Synchrony)
    iFC=zeros(N_areas);
    for n=1:N_areas
        for p=1:N_areas
            iFC(n,p)=cos(Phase_BOLD(n,t)-Phase_BOLD(p,t));
        end
    end
    
    % Get the leading eigenvector
    
    [V1,~]=eigs(iFC,1);
    % Make sure the largest component is negative
    if mean(V1>0)>.5
        V1=-V1;
    elseif mean(V1>0)==.5 && sum(V1(V1>0))>-sum(V1(V1<0))
        V1=-V1;
    end
    % Save V1 from all frames in all fMRI sessions in Leading eig
    Leading_Eig(:,t)=V1; %vertcat(V1,V2);
end
end

function entropy = tpm_entropy(IDX,datasizes)
% More or less copy-pasted from https://github.com/decolab/pnas-neuromod/blob/master/LEiDA_PsiloData.m

K = numel(unique(IDX(IDX>0)));

tpts_traversed = 1;
% entropy = nan(length(datasizes),1);
for ses = 1:length(datasizes)
    T = tpts_traversed:tpts_traversed+datasizes(ses)-1;
    
    Ctime=IDX(T);
    
    PTRANSITION=zeros(K,K);
    for c1=1:K
        for c2=1:K
            sumatr=0;
            for t=1:length(Ctime)-1
                if Ctime(t)==c1 && Ctime(t+1)==c2
                    sumatr=sumatr+1;
                end
            end
            if ~isempty((find(Ctime(1:length(Ctime)-1)==c1)))
                PTRANSITION(c1,c2)=sumatr/(length(find(Ctime(1:length(Ctime)-1)==c1)));
            end
        end
    end
    
    entropy{ses} = EntropyMarkov(PTRANSITION);
    
    tpts_traversed = tpts_traversed + datasizes(ses);
end
end

function H = EntropyMarkov(P)
number_states=size(P,1);
[V,D] = eig( P' );    % eigendecomp of P' (P=transition matrix)
[~,ii]= max(diag(D)) ;
st = V(:,ii);
p1 = abs(st)/sum(abs(st)); % p(i) (stationary!)

% Remark: you could also get an estimate of p(i) directly from the observations
% p(i) = sum( state == i )/T;

% Markov entropy:
Hi = zeros(1,number_states);
for row=1:number_states
    Hi(row) = -p1(row)*sum( (P(row,:)+eps).*log2(P(row,:)+eps) );
end

H = sum(Hi)/log2(number_states);
end

File: copbet_py/functions/CopBET_LEiDA_transition_entropy.py
"""
CopBET_LEiDA_transition_entropy
================================
Copenhagen Brain Entropy Toolbox: LEiDA transition entropy.

Evaluates the leading eigenvectors of instantaneous phase-coherence matrices
(via Hilbert transform), clusters them with K-means (squared Euclidean
distance), and computes the Markov transition rate entropy from the
resulting state sequence, following Kringelbach et al., 2020.

Input
-----
sessions : list of np.ndarray, each shape (T, N)
    ROI time series per session.
K : int
    Number of clusters for K-means.

Returns
-------
entropy : np.ndarray, shape (n_sessions,)
    Markov entropy per session (normalized by log2(K)).

Reference
---------
Kringelbach et al., 2020. "Dynamic coupling of whole-brain neuronal and
neurotransmitter systems."

Please cite McCulloch, Olsen et al., 2023 if you use CopBET.
"""

import numpy as np
from scipy.signal import hilbert
from sklearn.cluster import KMeans


def CopBET_LEiDA_transition_entropy(sessions, K=3, n_replicates=200, random_state=42):
    """
    Compute LEiDA Markov transition entropy.

    Parameters
    ----------
    sessions : list of np.ndarray, each shape (T, N)
    K : int
        Number of LEiDA states.
    n_replicates : int
        Number of K-means restarts.
    random_state : int

    Returns
    -------
    entropy : np.ndarray, shape (n_sessions,)
    """
    if K < 2:
        raise ValueError("K must be at least 2 for meaningful transition entropy.")
    # Compute leading eigenvectors for all sessions
    datasizes = []
    leading_eigs_list = []

    for ts in sessions:
        ts = np.asarray(ts, dtype=float)
        ts = ts - ts.mean(axis=0)
        eigs = _LEiDA(ts)          # (T, N)
        leading_eigs_list.append(eigs)
        datasizes.append(eigs.shape[0])

    Lead_eigs_all = np.vstack(leading_eigs_list)  # (T_total, N)

    # K-means on concatenated leading eigenvectors
    best_inertia = np.inf
    best_labels = None

    for seed in range(n_replicates):
        print(f"K-means replicate {seed + 1}/{n_replicates}...", end='\r')
        km = KMeans(n_clusters=K, n_init=1, max_iter=1000,
                    random_state=random_state + seed)
        km.fit(Lead_eigs_all)
        if km.inertia_ < best_inertia:
            best_inertia = km.inertia_
            best_labels = km.labels_

    # Compute Markov transition entropy per session
    entropy = np.full(len(sessions), np.nan)
    idx = 0
    for ses, size in enumerate(datasizes):
        seg = best_labels[idx:idx + size]
        P = _transition_matrix(seg, K)
        entropy[ses] = _markov_entropy(P, K)
        idx += size

    return entropy


def _LEiDA(ts):
    """
    Compute leading eigenvectors of instantaneous phase-coherence matrices.

    Parameters
    ----------
    ts : np.ndarray, shape (T, N)

    Returns
    -------
    leading_eigs : np.ndarray, shape (T, N)
    """
    T, N = ts.shape

    # Hilbert phase per region
    phase = np.angle(hilbert(ts, axis=0))  # (T, N)

    # via svd
    phase_stack = np.stack((np.cos(phase), np.sin(phase)), axis=-1)
    U, S, Vt = np.linalg.svd(phase_stack, full_matrices=False)
    leading_eigs = U[:,:,0]  # (N,)

    return leading_eigs


def _transition_matrix(labels, K):
    """
    Compute K×K Markov transition probability matrix from a state sequence.

    Parameters
    ----------
    labels : array-like, shape (T,)
        Integer state labels 0..K-1.
    K : int

    Returns
    -------
    P : np.ndarray, shape (K, K)
    """
    P = np.zeros((K, K))
    for t in range(len(labels) - 1):
        P[labels[t], labels[t + 1]] += 1

    # Normalize rows
    row_sums = P.sum(axis=1, keepdims=True)
    with np.errstate(invalid='ignore', divide='ignore'):
        P = np.where(row_sums > 0, P / row_sums, 0.0)

    return P


def _markov_entropy(P, K):
    """
    Compute Markov entropy rate from a transition matrix.

    Uses the stationary distribution as the weighting over rows.

    Parameters
    ----------
    P : np.ndarray, shape (K, K)
    K : int

    Returns
    -------
    H : float
        Normalized Markov entropy rate.
    """
    # Stationary distribution via left eigenvector (eigenvalue = 1)
    eigvals, eigvecs = np.linalg.eig(P.T)
    # Find eigenvector corresponding to eigenvalue closest to 1
    idx = np.argmin(np.abs(eigvals - 1.0))
    stat = np.abs(eigvecs[:, idx])
    stat = stat / stat.sum()

    eps = np.finfo(float).eps
    Hi = np.zeros(K)
    for row in range(K):
        row_p = P[row, :] + eps
        Hi[row] = -stat[row] * np.sum(row_p * np.log2(row_p))

    H = np.sum(Hi) / np.log2(K)

    return float(H)
